Ivy Frontend Instance Method Setup

jkeane508 opened the pull request.

jkeane508 commented on ivy_tests/test_ivy/helpers.py:
```diff
@@ -1756,15 +1757,33 @@ def test_frontend_function(
+                fn=instance.__getattribute__(function_name), args=args, kwargs=kwargs
+            )
+
+        setattr(instance, "ivy_instance_method", frontend_fn)
```
Issue here, can't figure out how to call the instance method for the instance array.

jkeane508 commented on ivy_tests/test_ivy/helpers.py:
```diff
@@ -1716,10 +1717,10 @@ def test_frontend_function(
 
     # parse function name and frontend submodules (i.e. jax.lax, jax.numpy etc.)
-    *frontend_submods, fn_tree = fn_tree.split(".")
+    *frontend_submods, function_name = fn_tree.split(".")
```
Changes this to function name for readability

jkeane508 reviewed (commented):
Hey @VedPatwardhan,

Wondering if I could get your input on the testing as I add behvaiour for frontend tests to handle instance methods

VedPatwardhan reviewed (commented):
Hi @jkeane508, thanks for requesting the review. Even though I don't have the exact idea of what you're trying to do, I think for adding instance methods, you'll first need classes named `ndarray`, `Tensor`, etc. and they would then be extending the `ivy.Array` class. Inside that class, you would then define these functions as instance method, such that they use the functions in the API of `ivy.functional.frontends.<fw>` internally, passing `self` as the argument, similar to how we add the instance methods for arrays and containers in the main functional API. 
And then for adding the tests, we would just do the exact same thing we did with the functional tests, i.e. create an instance and then call `__getattribute__` on it.

Again, as I said before, I'm not 100% sure if this is how it should be done. @djl11 could you please confirm if this is the right way? Thanks 🙂

jkeane508 commented:
Hey @VedPatwardhan,

Thanks for you input! I had thought extending the `ivy.array` class was what was needed. I had just thought that there would be away around it without cluttering the number of attributes attached to it. I understand now that I will need to extend the framework specific array classes and the ivy array class for all the instance methods.

jkeane508 commented:
Hey @VedPatwardhan,

I've made some further changes to this PR. Basically what I've done is create a general frontend_array class based off the original ivy.array class and used this create framework specific subclasses that (intend to) mimic these framework specific classes.

Have a look and tell me what you think about this idea :)

PS. Ive only semi-implemented the Jax version

VedPatwardhan reviewed (commented):
Hi @jkeane508, sorry about the delayed response. I'm not exactly sure if this is the right way to do it, because I thought we could just extend the `ivy.Array` class in the frontend_array class and then just add the additional methods for the frontends  which aren't present in `ivy.Array`. So I think I should request Dan's review on this one.

jkeane508 commented:
Hi @VedPatwardhan,

I know realise that I have grossly overcomplicated this and the easiest way to do this is to create a class which simply stores the array as an attribute and call the instance methods on top.

bicycleman15 reviewed (commented):
Hey @jkeane508, the overall design seems good (as well as simple which is a good thing), but adding onto @VedPatwardhan point, I still think inheriting from `ivy.Array` would still be a better option. 

However, I am not sure if let's say some framework does not have some function `x` as instance method, but `ivy.Array` does have function `x` as instance method, we would be adding extra instance methods to this new frontend array class we are creating. Maybe we could blacklist these functions `x` or something ?

jkeane508 commented on ivy_tests/test_ivy/test_frontends/test_jax/test_jax_devicearray.py:
```diff
@@ -0,0 +1,57 @@
+    fw,
+):
+    x, dtype, shape, dimensions = x_reshape_permut
+    helpers.test_frontend_function(
```
The testing is still very much wip, Im looking into that today :)

djl11 replied:
we will need to either create a new helper function for instance methods on the frontend arrays or extend `test_frontend_function` to support this. One approach would be to add something like `test_frontend_method` which mirrors `test_method`, and then we create a frontend array instance inside the helper function and then test the specified method. However, it might be easier to simply extend `test_frontend_function` with a flag for converting to frontend array instances. Will have a think and assign a task for this soon.

jkeane508 commented:
> Hey @jkeane508, the overall design seems good (as well as simple which is a good thing), but adding onto @VedPatwardhan point, I still think inheriting from `ivy.Array` would still be a better option.
> 
> However, I am not sure if let's say some framework does not have some function `x` as instance method, but `ivy.Array` does have function `x` as instance method, we would be adding extra instance methods to this new frontend array class we are creating. Maybe we could blacklist these functions `x` or something ?

Yea, so I think what's easier is just to write a class which wraps all of the array classes (including ivy), and stores the array in a `data` attribute. The specific frontend class will store the instance methods and you would then have the instance methods apply the functions to `x.data` in the return. This is essentially a very basic idea of what the ivy array class does (it just has more bells and whistles')

If we inherit the ivy.array class in the frontend arrays, we are essentially extending the ivy array class functionality which means that we have to call essentially all of the Ivy functional API every time a frontend array is instantiated which Im sure is not ideal.

djl11 reviewed (changes requested):
small changes requested, and the helper function is pending which will hold up getting this wip PR merged.

jkeane508 commented:
@djl11,

Makes sense, I think extending the frontend test helper would be cohesive with the backend testing, but may slow down testing in general with the addition of code.

Possibly a helper for the frontend test helper that when triggered returns all the necessary outputs to run the value testing in the original frontend test helper?

jkeane508 commented:
Hi @VedPatwardhan, @djl11 and @bicycleman15 :)

I've implemented a version of the helper for frontend instance method testing, and the test examples.

Currently I'm having trouble with getting the frontend classes into the correct namespace particularly the ndarray class from numpy. This is because its nested within the ndarray module currently.

djl11 commented:
> Hi @VedPatwardhan, @djl11 and @bicycleman15 :)
> 
> I've implemented a version of the helper for frontend instance method testing, and the test examples.
> 
> Currently I'm having trouble with getting the NumPy frontend as it has a ndarray module already, Im not sure how to structure it such that ndarray can be a reference for the class and the module?

Hi James, could you give more context (with the inclusion of links to lines of code etc.) to explain this issue with NumPy, I wasn't able to follow what you're referring to on first read. Thanks!

jkeane508 commented:
> > Hi @VedPatwardhan, @djl11 and @bicycleman15 :)
> > I've implemented a version of the helper for frontend instance method testing, and the test examples.
> > Currently I'm having trouble with getting the NumPy frontend as it has a ndarray module already, Im not sure how to structure it such that ndarray can be a reference for the class and the module?
> 
> Hey @jkeane508, could you give more context (with the inclusion of links to lines of code etc.) to explain this issue with NumPy, I wasn't able to follow what you're referring to on first read. Thanks!

Oh yea sorry, basically had an issue with the init files and calling `frontends.numpy.ndarray` as it could either be the class or the module. But I have fixed it now I think :)

djl11 reviewed (changes requested):
One small change requested, otherwise it looks good. The rational for this change is that we should handle `ivy.Array` instances behind the scenes for the frontend array class. This opens up the possibility to use `ivy.Array` instance methods on `self.data` and also perform call operators on `self.data` (`+`,`-`, `*`, `/` etc.) in a unified way, which would not be the case if `self.data` was an `ivy.NativeArray` instance. Following these small changes, as long as the unit tests are passing, and when stepping through the test helper you are confident that it's testing for the right thing, then I think it all looks great!

jkeane508 commented on ivy/functional/frontends/jax/devicearray.py:
```diff
@@ -0,0 +1,20 @@
+class DeviceArray:
+    def __init__(self, data):
+        if ivy.is_ivy_array(data):
+            self.data = data.data
```
Yes, I came across this issue when lists where passed in. Its been changed locally :)
